From: Paolo Bonzini <pbonzini@redhat.com>
Date: Thu, 18 Nov 2010 18:18:04 -0500
Subject: [xen] hvm: add HVMOP_get_time hypercall
Message-id: <1290104284-29836-1-git-send-email-pbonzini@redhat.com>
Patchwork-id: 29497
O-Subject: [RHEL5.6 Xen PATCH] hvm: Add HVMOP_get_time hypercall
Bugzilla: 638082
RH-Acked-by: Laszlo Ersek <lersek@redhat.com>

Bugzilla: 638082

Upstream status: http://xenbits.xensource.com/xen-unstable.hg/rev/21772

Brew build: https://brewweb.devel.redhat.com/taskinfo?taskID=2901669

This hypercall is necessary to use SCHEDOP_poll from HVM guests.  I have
plans to use it in the Windows PV drivers, and it would be nice to have it
in the 5.6 hypervisor since the backport is small and safe.

---
 arch/x86/hvm/hvm.c          |    9 +++++++++
 include/public/hvm/hvm_op.h |    8 ++++++++
 2 files changed, 17 insertions(+), 0 deletions(-)

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/x86/hvm/hvm.c b/arch/x86/hvm/hvm.c
index a3ccaba..a84620b 100644
--- a/arch/x86/hvm/hvm.c
+++ b/arch/x86/hvm/hvm.c
@@ -1582,6 +1582,15 @@ long do_hvm_op(unsigned long op, XEN_GUEST_HANDLE(void) arg)
         rc = guest_handle_is_null(arg) ? hvmop_flush_tlb_all() : -ENOSYS;
         break;
 
+    case HVMOP_get_time: {
+        xen_hvm_get_time_t gxt;
+
+        gxt.now = NOW();
+        if ( copy_to_guest(arg, &gxt, 1) )
+            rc = -EFAULT;
+        break;
+    }
+
     default:
     {
         gdprintk(XENLOG_WARNING, "Bad HVM op %ld.\n", op);
diff --git a/include/public/hvm/hvm_op.h b/include/public/hvm/hvm_op.h
index b21b0f7..8c47d3d 100644
--- a/include/public/hvm/hvm_op.h
+++ b/include/public/hvm/hvm_op.h
@@ -73,4 +73,12 @@ DEFINE_XEN_GUEST_HANDLE(xen_hvm_set_pci_link_route_t);
 /* Flushes all VCPU TLBs: @arg must be NULL. */
 #define HVMOP_flush_tlbs          5
 
+/* Get the current Xen time, in nanoseconds since system boot. */
+#define HVMOP_get_time              10
+struct xen_hvm_get_time {
+    uint64_t now;      /* OUT */
+};
+typedef struct xen_hvm_get_time xen_hvm_get_time_t;
+DEFINE_XEN_GUEST_HANDLE(xen_hvm_get_time_t);
+
 #endif /* __XEN_PUBLIC_HVM_HVM_OP_H__ */
